File: solutions/4192-minimum-cost-to-partition-a-binary-string/minimum-cost-to-partition-a-binary-string.java
class Solution {
    public long minCost(String s, int encCost, int flatCost) {
        int n_size = s.length();
        long[] pre = new long[n_size];
        pre[0] = (s.charAt(0) == '1') ? 1 : 0;
        for (int i = 1; i < n_size; i++) {
            pre[i] = pre[i - 1] + (s.charAt(i) == '1' ? 1 : 0);
        }
        
        return cost(0, n_size - 1, s, encCost, flatCost, pre);
    }

    private long cost(int l, int r, String s, int ec, int fc, long[] pre) {
        if (l > r) return (long) 1e18;
        
        long count = (l > 0) ? pre[r] - pre[l - 1] : pre[r];
        long keep = fc;
        long n = r - l + 1;
        
        if (count > 0) {
            keep = n * count * (long) ec;
        }
        
        long rcost = (long) 1e18;
        long lcost = (long) 1e18;
        
        if (n % 2 == 0) {
            int mid = l + (r - l) / 2;
            lcost = cost(l, mid, s, ec, fc, pre);
            rcost = cost(mid + 1, r, s, ec, fc, pre);
        }
        
        return Math.min(keep, rcost + lcost);
    }
}

File: solutions/4192-minimum-cost-to-partition-a-binary-string/minimum-cost-to-partition-a-binary-string.py
class Solution:
    def solve(self, s, encCost, flatCost, start, end):
        l = end - start + 1
        
        count = 0
        for i in range(start, end + 1):
            if s[i] == '1':
                count += 1
        
        if count:
            cost = encCost * l * count
        else:
            cost = flatCost
        
        mid = (start + end) // 2
        
        if l % 2 == 0:
            cost1 = self.solve(s, encCost, flatCost, start, mid)
            cost2 = self.solve(s, encCost, flatCost, mid + 1, end)
            cost = min(cost, cost1 + cost2)
        
        return cost

    def minCost(self, s: str, encCost: int, flatCost: int) -> int:
        return self.solve(s, encCost, flatCost, 0, len(s) - 1)